src/azul/service/repository_controller.py

```diff
@@ -258,11 +258,14 @@ def list_sources() -> Iterable[SourceRef]:
                 raise UnauthorizedError
 
         if authentication is None:
-            # FIXME: Determine public snapshots
-            #        https://github.com/DataBiosphere/azul/issues/2978
+            authentication, authentication_identity = plugin.default_authentication()
```

Comment:
Nah, this is back to being being overly complicated. If the issue is not wanting to create many cache lines for the public SA, then we should find a more comprehensible way of addressing that.

Maybe the fall-back to the public service account should be done in the list_sources implementation of the TDR plugin. The cache key should be None when authentication is None and authentication.identity() when it isn't. The default_authentication method would go away in that approach.

Reply:
Looks like you haven't addressed this yet, which is fine since you haven't asked for review. We do have to accelerate this a bit so let me be more specific and lay this out in a patch:

```diff
Index: src/azul/plugins/repository/tdr/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr/__init__.py b/src/azul/plugins/repository/tdr/__init__.py
--- a/src/azul/plugins/repository/tdr/__init__.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/plugins/repository/tdr/__init__.py	(date 1626193217546)
@@ -179,29 +179,27 @@
     def sources(self) -> AbstractSet[TDRSourceSpec]:
         return self._sources
 
-    def default_authentication(self) -> Tuple[Authentication, str]:
-        public_tdr = self._public_authenticated_tdr()
-        return OAuth2(public_tdr.credentials.token), public_tdr.authentication_id()
-
     def list_sources(self,
                      authentication: Optional[Authentication]
                      ) -> List[TDRSourceRef]:
-        if isinstance(authentication, OAuth2):
+        if authentication is None:
+            tdr = TDRClient.with_public_service_account_credentials()
+        elif isinstance(authentication, OAuth2):
             tdr = TDRClient.with_user_credentials(authentication)
-            configured_specs_by_name = {spec.name: spec for spec in self.sources}
-            snapshot_ids_by_name = {
-                name: id
-                for id, name in tdr.snapshot_names_by_id().items()
-                if name in configured_specs_by_name
-            }
-            return [
-                TDRSourceRef(id=id,
-                             spec=configured_specs_by_name[name])
-                for name, id in snapshot_ids_by_name.items()
-            ]
-        else:
-            raise PermissionError('Unsupported authentication format',
-                                  type(authentication))
+        else:
+            raise PermissionError('Unsupported authentication format',
+                                  type(authentication))
+        configured_specs_by_name = {spec.name: spec for spec in self.sources}
+        snapshot_ids_by_name = {
+            name: id
+            for id, name in tdr.snapshot_names_by_id().items()
+            if name in configured_specs_by_name
+        }
+        return [
+            TDRSourceRef(id=id,
+                         spec=configured_specs_by_name[name])
+            for name, id in snapshot_ids_by_name.items()
+        ]
 
     @property
     def tdr(self):
Index: src/azul/plugins/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/__init__.py b/src/azul/plugins/__init__.py
--- a/src/azul/plugins/__init__.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/plugins/__init__.py	(date 1626194964664)
@@ -196,14 +196,6 @@
         """
         raise NotImplementedError
 
-    def default_authentication(self) -> Union[Tuple[Authentication, str],
-                                              Tuple[None, None]]:
-        """
-        Authentication and authentication identity to use for accessing the
-        remote repository when none is provided by the request.
-        """
-        return None, None
-
     @abstractmethod
     def list_sources(self,
                      authentication: Optional[Authentication]
Index: src/azul/service/repository_controller.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/repository_controller.py b/src/azul/service/repository_controller.py
--- a/src/azul/service/repository_controller.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/service/repository_controller.py	(date 1626193266346)
@@ -251,32 +251,28 @@
         authentication = request.authentication
         plugin = self.repository_plugin(catalog)
 
-        def list_sources() -> Iterable[SourceRef]:
+        cache_key = (
+            catalog,
+            '' if authentication is None else authentication.identity()
+        )
+        joiner = ':'
+        assert not any(joiner in c for c in cache_key), cache_key
+        cache_key = joiner.join(cache_key)
+        try:
+            sources = self._source_cache_service.get(cache_key)
+        except CacheMiss:
             try:
-                return plugin.list_sources(authentication)
+                sources = plugin.list_sources(authentication)
             except PermissionError:
                 raise UnauthorizedError
-
-        if authentication is None:
-            authentication, authentication_identity = plugin.default_authentication()
-        else:
-            authentication_identity = authentication.identity()
-
-        if authentication is None:
-            sources = list_sources()
+            else:
+                sources = [source.to_json() for source in sources]
+                self._source_cache_service.put(cache_key, sources)
         else:
-            cache_key = f'{catalog}:{authentication_identity}'
-            try:
-                source_jsons = self._source_cache_service.get(cache_key)
-            except CacheMiss:
-                sources = list_sources()
-                source_jsons = [source.to_json() for source in sources]
-                self._source_cache_service.put(cache_key, source_jsons)
-            else:
-                sources = [
-                    plugin.source_from_json(source_json)
-                    for source_json in source_jsons
-                ]
+            sources = [
+                plugin.source_from_json(source)
+                for source in sources
+            ]
         return [
             {
                 'sourceId': source.id,
```

Aside from simplifying things, this also removes `source_jsons` and reuses `sources` instead, and has more robust composition of the cache key.

It caches DSS sources but I see no downsides to that. It also wouldn't invalidate the "public" cache line if the public service account changes but that is rare. The current implementation would still let the "public" cache go stale if the permissions of the SA change. The latter is much more likely to occur and we address that by having a short-ish expiration. The short expiration also takes care of a change in the public SA.